// @rss3/api-core/dist/dsl.cjs
'use strict';

var requests = require('./requests');



Object.defineProperty(exports, "getActivities", {
  enumerable: true,
  get: function () { return requests.getActivities; }
});
Object.defineProperty(exports, "getActivitiesByAccounts", {
  enumerable: true,
  get: function () { return requests.getActivitiesByAccounts; }
});
Object.defineProperty(exports, "getActivity", {
  enumerable: true,
  get: function () { return requests.getActivity; }
});
Object.defineProperty(exports, "getFederatedActivitiesByAccount", {
  enumerable: true,
  get: function () { return requests.getFederatedActivitiesByAccount; }
});
Object.defineProperty(exports, "getFederatedActivitiesByAccounts", {
  enumerable: true,
  get: function () { return requests.getFederatedActivitiesByAccounts; }
});
Object.defineProperty(exports, "getFederatedActivity", {
  enumerable: true,
  get: function () { return requests.getFederatedActivity; }
});
Object.defineProperty(exports, "getFederatedNetworkActivities", {
  enumerable: true,
  get: function () { return requests.getFederatedNetworkActivities; }
});
Object.defineProperty(exports, "getFederatedPlatformActivities", {
  enumerable: true,
  get: function () { return requests.getFederatedPlatformActivities; }
});
Object.defineProperty(exports, "getNetworkActivities", {
  enumerable: true,
  get: function () { return requests.getNetworkActivities; }
});
Object.defineProperty(exports, "getPlatformActivities", {
  enumerable: true,
  get: function () { return requests.getPlatformActivities; }
});
Object.defineProperty(exports, "getRSSActivity", {
  enumerable: true,
  get: function () { return requests.getRSSActivity; }
});
