// @rss3/api-core/dist/nta.cjs
'use strict';

var requests = require('./requests');



Object.defineProperty(exports, "getAllNodes", {
  enumerable: true,
  get: function () { return requests.getAllNodes; }
});
Object.defineProperty(exports, "getAverageEpochsApy", {
  enumerable: true,
  get: function () { return requests.getAverageEpochsApy; }
});
Object.defineProperty(exports, "getBridgingTransaction", {
  enumerable: true,
  get: function () { return requests.getBridgingTransaction; }
});
Object.defineProperty(exports, "getBridgingTransactions", {
  enumerable: true,
  get: function () { return requests.getBridgingTransactions; }
});
Object.defineProperty(exports, "getChip", {
  enumerable: true,
  get: function () { return requests.getChip; }
});
Object.defineProperty(exports, "getChipSvg", {
  enumerable: true,
  get: function () { return requests.getChipSvg; }
});
Object.defineProperty(exports, "getChipSvgSrc", {
  enumerable: true,
  get: function () { return requests.getChipSvgSrc; }
});
Object.defineProperty(exports, "getChipSvgURL", {
  enumerable: true,
  get: function () { return requests.getChipSvgURL; }
});
Object.defineProperty(exports, "getChips", {
  enumerable: true,
  get: function () { return requests.getChips; }
});
Object.defineProperty(exports, "getEpoch", {
  enumerable: true,
  get: function () { return requests.getEpoch; }
});
Object.defineProperty(exports, "getEpochTransaction", {
  enumerable: true,
  get: function () { return requests.getEpochTransaction; }
});
Object.defineProperty(exports, "getEpochs", {
  enumerable: true,
  get: function () { return requests.getEpochs; }
});
Object.defineProperty(exports, "getEpochsApySnapshot", {
  enumerable: true,
  get: function () { return requests.getEpochsApySnapshot; }
});
Object.defineProperty(exports, "getNetworkAssets", {
  enumerable: true,
  get: function () { return requests.getNetworkAssets; }
});
Object.defineProperty(exports, "getNetworkConfig", {
  enumerable: true,
  get: function () { return requests.getNetworkConfig; }
});
Object.defineProperty(exports, "getNode", {
  enumerable: true,
  get: function () { return requests.getNode; }
});
Object.defineProperty(exports, "getNodeAvatarSvg", {
  enumerable: true,
  get: function () { return requests.getNodeAvatarSvg; }
});
Object.defineProperty(exports, "getNodeAvatarSvgSrc", {
  enumerable: true,
  get: function () { return requests.getNodeAvatarSvgSrc; }
});
Object.defineProperty(exports, "getNodeAvatarSvgURL", {
  enumerable: true,
  get: function () { return requests.getNodeAvatarSvgURL; }
});
Object.defineProperty(exports, "getNodeCountSnapshot", {
  enumerable: true,
  get: function () { return requests.getNodeCountSnapshot; }
});
Object.defineProperty(exports, "getNodeEvents", {
  enumerable: true,
  get: function () { return requests.getNodeEvents; }
});
Object.defineProperty(exports, "getNodeOperationProfit", {
  enumerable: true,
  get: function () { return requests.getNodeOperationProfit; }
});
Object.defineProperty(exports, "getNodeRewards", {
  enumerable: true,
  get: function () { return requests.getNodeRewards; }
});
Object.defineProperty(exports, "getOperationProfitSnapshot", {
  enumerable: true,
  get: function () { return requests.getOperationProfitSnapshot; }
});
Object.defineProperty(exports, "getStakerCountSnapshot", {
  enumerable: true,
  get: function () { return requests.getStakerCountSnapshot; }
});
Object.defineProperty(exports, "getStakerProfitSnapshot", {
  enumerable: true,
  get: function () { return requests.getStakerProfitSnapshot; }
});
Object.defineProperty(exports, "getStakingProfit", {
  enumerable: true,
  get: function () { return requests.getStakingProfit; }
});
Object.defineProperty(exports, "getStakingStat", {
  enumerable: true,
  get: function () { return requests.getStakingStat; }
});
Object.defineProperty(exports, "getStakingTransaction", {
  enumerable: true,
  get: function () { return requests.getStakingTransaction; }
});
Object.defineProperty(exports, "getStakingTransactions", {
  enumerable: true,
  get: function () { return requests.getStakingTransactions; }
});
Object.defineProperty(exports, "getStakings", {
  enumerable: true,
  get: function () { return requests.getStakings; }
});
Object.defineProperty(exports, "getTokenSupply", {
  enumerable: true,
  get: function () { return requests.getTokenSupply; }
});
Object.defineProperty(exports, "getTokenTvl", {
  enumerable: true,
  get: function () { return requests.getTokenTvl; }
});
Object.defineProperty(exports, "getTotalRequests", {
  enumerable: true,
  get: function () { return requests.getTotalRequests; }
});
