// @rss3/api-utils/dist/readable/content/index.global.js
var DslApiUtils = (function (exports) {
	'use strict';

	function l(e){return `${e.tag}-${e.type}`}function s(e,r){r[l(e)]?.(e.metadata);}function u(e){let[r]=e.actions??[];return r?c(e,r):void 0}function c(e,r){let o;return s(r,{"social-post":t=>{o=n(e,r,t);},"social-comment":t=>{o=n(e,r,t);},"social-mint":t=>{o=n(e,r,t);},"social-share":t=>{o=n(e,r,t);},"social-revise":t=>{o=n(e,r,t);},"social-delete":t=>{o=n(e,r,t);}}),o}function a(e,r){if(!e)return e;let o=e.replaceAll("\u2026","");return r?.startsWith(o)?void 0:o}function m(e){return e?!!(e.body||e.media||e.title):!1}function n(e,r,o){let t=o.target,i=t?{authorUrl:t.authorUrl,handle:t.handle,address:r.to,profileId:t.profileId,title:a(t.title,t.body),body:t.body,media:t.media}:void 0;i?.media&&e.network?.toLowerCase()==="mastodon"&&(i.media=i.media.slice(1));let d={authorUrl:o.authorUrl,handle:o.handle,address:r.from,profileId:o.profileId,title:a(o.title,o.body),body:o.body,media:o.media,target:i};return d.media&&e.network?.toLowerCase()==="mastodon"&&(d.media=d.media.slice(1)),d}

	exports.checkTargetExist = m;
	exports.extractContent = c;
	exports.formatContent = u;
	exports.formatTitle = a;

	return exports;

})({});
