// @rss3/api-utils/dist/readable/activity/token.global.js
var DslApiUtils = (function (exports) {
	'use strict';

	function l(e,n){let r="0.00",a,o=e;if(o){if(Number.isNaN(Number(o))||Number(o)===0)return "0.00";Number(o)>0&&n!==void 0&&(o=s(o,n));let i=o.match(/^(0.0+)(.*)/);Number(o)>0&&Number(o)<1?a=i?i[1]+i[2].slice(0,3):`0.${o.split("0.")[1]?.slice(0,3)??"0"}`:a=Number(Number(o).toFixed(3)).toString(),a.includes(".")||(a=Number(o).toFixed(2)),r=a.replace(/(\d)(?=(\d{3})+\.\d+)/g,"$1,");}return r}function s(e,n){return (e.length>n?`${e.slice(0,e.length-n)}.${e.slice(-n)}`:`0.${"0".repeat(n-e.length)}${e}`).replace(/0+$/,"")}function t(e,n="",r){return {type:e,content:n,meta:r}}var m=u(" "),c=t("separator","; ");function g(e,n=m){return e.reduce((r,a)=>(r.push(n,a),r),[]).slice(1)}function u(e){return t("text",e)}function k(e){return t("name",e)}function b(e){return t("image",e||"")}function x(e){return t("network",e||"")}function y(e){return t("time",new Date(e*1e3).toJSON())}function T(e){return t("address",e||"")}function h(e,n,r){return n?t("address",e||"",{address:n,network:r||""}):t("address",e||"")}var d="115792089237316195423570985008687907853269984665640564039457584007913129639935";function N(e){return e?e.value===d?[t("number","infinite"),t("symbol",e.symbol)]:[t("symbolImage",e.image),t("number",l(e.value,e.decimals)||"0"),t("symbol",e.symbol)]:[t("number","0")]}function w(e){let n="";return e.platform?(n=e.platform,[u("on"),t("platform",n)]):[]}function A(e){if(e.tag!=="social")return u("");if(e.metadata){if("title"in e.metadata&&e.metadata.title&&!(e.platform==="Lens"&&e.metadata.title))return t("html",e.metadata.title);if("body"in e.metadata&&e.metadata.body)return t("html",e.metadata.body);if("target"in e.metadata&&e.metadata.target&&"body"in e.metadata.target&&e.metadata.target.body)return t("html",e.metadata.target.body)}return t("html","")}function I(e,n){let r=n?.preview;return r?[t("assetImage",r),t("name",e,n)]:[t("name",e,n)]}

	exports.join = g;
	exports.token = t;
	exports.tokenAddr = T;
	exports.tokenAsset = I;
	exports.tokenHandle = h;
	exports.tokenImage = b;
	exports.tokenName = k;
	exports.tokenNetwork = x;
	exports.tokenPlatform = w;
	exports.tokenPost = A;
	exports.tokenSeparator = c;
	exports.tokenSpace = m;
	exports.tokenText = u;
	exports.tokenTime = y;
	exports.tokenValue = N;

	return exports;

})({});
